File: innomat/innomat/page/crm_dashboard/crm_dashboard.py
# -*- coding: utf-8 -*-
# Copyright (c) 2026, Innomat, Asprotec and libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
import json
from frappe import _
from six import string_types
from frappe.utils import getdate, nowdate

EDITABLE_FIELDS = ["communication_type", "completed", "date", "time", "duration",
                   "user", "preparation", "note", "follow_up"]


@frappe.whitelist()
def get_upcoming(user=None, limit=25):
    """Return the open (not completed) communication entries per type."""
    frappe.has_permission("Lead", "read", throw=True)
    complete_past_entries()

    limit = min(int(limit or 25), 100)
    conditions = ""
    values = {"limit": limit}
    if user:
        conditions = " AND k.`user` = %(user)s"
        values["user"] = user

    data = {}
    for communication_type in ["Telefon", "Remotebesprechung", "Vor Ort Besprechung"]:
        values["communication_type"] = communication_type
        data[communication_type] = frappe.db.sql("""
            SELECT
                k.`name` AS `name`,
                k.`parent` AS `lead`,
                k.`communication_type` AS `communication_type`,
                k.`date` AS `date`,
                k.`time` AS `time`,
                k.`duration` AS `duration`,
                k.`user` AS `user`,
                k.`preparation` AS `preparation`,
                k.`note` AS `note`,
                k.`follow_up` AS `follow_up`,
                l.`lead_name` AS `lead_name`,
                l.`company_name` AS `company_name`
            FROM `tabLead Kommunikation` AS k
            LEFT JOIN `tabLead` AS l ON l.`name` = k.`parent`
            WHERE k.`parenttype` = 'Lead'
              AND k.`communication_type` = %(communication_type)s
              AND IFNULL(k.`completed`, 0) = 0
              {conditions}
            ORDER BY k.`date` ASC, k.`time` ASC
            LIMIT %(limit)s
        """.format(conditions=conditions), values, as_dict=True)

    return data


def complete_past_entries():
    """Complete open communication rows dated before today."""
    rows = frappe.db.sql("""
        SELECT k.`name`, k.`parent`
        FROM `tabLead Kommunikation` AS k
        WHERE k.`parenttype` = 'Lead'
          AND IFNULL(k.`completed`, 0) = 0
          AND k.`date` < CURDATE()
    """, as_dict=True)

    rows_by_lead = {}
    for row in rows:
        rows_by_lead.setdefault(row.parent, set()).add(row.name)

    completed = 0
    for parent, row_names in rows_by_lead.items():
        lead = frappe.get_doc("Lead", parent)
        lead.check_permission("write")
        for communication in lead.verlauf:
            if communication.name in row_names and not communication.completed:
                communication.completed = 1
                completed += 1
        if lead.contact_date and getdate(lead.contact_date) < getdate(nowdate()):
            lead.contact_date = None
        lead.save()

    if completed:
        frappe.db.commit()
    return completed


def normalize_communication_types():
    """Rename legacy communication types after the selection change."""
    frappe.db.sql("""
        UPDATE `tabLead Kommunikation`
        SET `communication_type` = 'Telefon'
        WHERE `communication_type` = 'Telefonat'
    """)
    frappe.db.sql("""
        UPDATE `tabLead Kommunikation`
        SET `communication_type` = 'Vor Ort Besprechung'
        WHERE `communication_type` = 'Besprechung'
    """)
    frappe.db.commit()


@frappe.whitelist()
def update_entry(name, values):
    """Update a single Lead Kommunikation row through its parent Lead."""
    if isinstance(values, string_types):
        values = json.loads(values)

    parent = frappe.db.get_value("Lead Kommunikation", name, "parent")
    if not parent:
        frappe.throw(_("Eintrag nicht gefunden"))

    lead = frappe.get_doc("Lead", parent)
    lead.check_permission("write")

    for row in lead.verlauf:
        if row.name == name:
            for field in EDITABLE_FIELDS:
                if field in values:
                    row.set(field, values.get(field))
            break
    else:
        frappe.throw(_("Eintrag nicht gefunden"))

    lead.save()
    frappe.db.commit()
    return name


@frappe.whitelist()
def create_follow_up_entry(name, values):
    """Create a new communication row for the same Lead and complete the old row."""
    if isinstance(values, string_types):
        values = json.loads(values)

    parent = frappe.db.get_value("Lead Kommunikation", name, "parent")
    if not parent:
        frappe.throw(_("Eintrag nicht gefunden"))

    if not values.get("date"):
        frappe.throw(_("Bitte ein Datum für den neuen Termin eingeben"))

    lead = frappe.get_doc("Lead", parent)
    lead.check_permission("write")

    current_row = None
    for row in lead.verlauf:
        if row.name == name:
            current_row = row
            break

    if not current_row:
        frappe.throw(_("Eintrag nicht gefunden"))

    current_row.completed = 1
    lead.append("verlauf", {
        "doctype": "Lead Kommunikation",
        "communication_type": values.get("communication_type") or current_row.communication_type or "Telefon",
        "completed": 0,
        "date": values.get("date"),
        "time": values.get("time"),
        "duration": values.get("duration") or current_row.duration or 1.0,
        "user": values.get("user") or current_row.user,
        "preparation": values.get("preparation"),
        "note": values.get("note"),
        "follow_up": values.get("follow_up")
    })
    lead.save()
    frappe.db.commit()
    return lead.name


@frappe.whitelist()
def complete_all(user=None):
    """Mark all visible open communication rows as completed."""
    frappe.has_permission("Lead", "read", throw=True)

    user_condition = ""
    values = {}
    if user:
        user_condition = " AND k.`user` = %(user)s"
        values["user"] = user

    rows = frappe.db.sql("""
        SELECT k.`name`, k.`parent`
        FROM `tabLead Kommunikation` AS k
        WHERE k.`parenttype` = 'Lead'
          AND IFNULL(k.`completed`, 0) = 0
          {user_condition}
    """.format(user_condition=user_condition), values, as_dict=True)

    rows_by_lead = {}
    for row in rows:
        rows_by_lead.setdefault(row.parent, set()).add(row.name)

    completed = 0
    for parent, row_names in rows_by_lead.items():
        lead = frappe.get_doc("Lead", parent)
        lead.check_permission("write")
        for communication in lead.verlauf:
            if communication.name in row_names and not communication.completed:
                communication.completed = 1
                completed += 1
        lead.save()

    frappe.db.commit()
    return completed


@frappe.whitelist()
def create_entry(values):
    """Create a communication row on an existing or newly created Lead."""
    if isinstance(values, string_types):
        values = json.loads(values)

    lead_name = values.get("lead")
    new_lead_name = (values.get("new_lead_name") or "").strip()
    if lead_name:
        lead = frappe.get_doc("Lead", lead_name)
        lead.check_permission("write")
    elif new_lead_name:
        frappe.has_permission("Lead", "create", throw=True)
        lead = frappe.get_doc({
            "doctype": "Lead",
            "lead_name": new_lead_name
        })
    else:
        frappe.throw(_("Bitte einen bestehenden Lead auswählen oder einen neuen Namen eingeben"))

    communication = {
        "doctype": "Lead Kommunikation",
        "communication_type": values.get("communication_type") or "Telefon",
        "completed": 0,
        "date": values.get("date"),
        "time": values.get("time"),
        "duration": values.get("duration") or 1.0,
        "user": values.get("user"),
        "preparation": values.get("preparation"),
        "note": values.get("note"),
        "follow_up": values.get("follow_up")
    }
    lead.append("verlauf", communication)
    lead.save()
    frappe.db.commit()
    return lead.name

File: innomat/innomat/page/crm_dashboard/crm_dashboard.js
frappe.pages['crm_dashboard'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __("CRM Dashboard"),
        single_column: true
    });

    frappe.crm_dashboard.make(page);
    frappe.crm_dashboard.refresh();

    frappe.breadcrumbs.add("Innomat");
};

frappe.crm_dashboard = {
    page: null,
    body: null,
    entries: {},

    make: function(page) {
        var me = frappe.crm_dashboard;
        me.page = page;

        me.user_field = page.add_field({
            fieldname: 'user',
            label: __("Benutzer"),
            fieldtype: 'Link',
            options: 'User',
            change: function() { me.refresh(); }
        });

        page.set_primary_action(__("Neuer Eintrag"), function() { me.new_entry(); }, 'add');
        page.add_inner_button(__("Aktualisieren"), function() { me.refresh(); }, 'refresh');

        me.body = $('<div class="crm-dashboard"></div>').appendTo(page.main);
        me.body.on('click', '.crm-entry', function(e) {
            if ($(e.target).closest('.crm-open-lead, .crm-create-follow-up').length) { return; }
            me.edit_entry($(this).attr('data-name'));
        });
        me.body.on('click', '.crm-open-lead', function(e) {
            e.preventDefault();
            e.stopPropagation();
            frappe.set_route("Form", "Lead", $(this).attr('data-lead'));
        });
        me.body.on('click', '.crm-create-follow-up', function(e) {
            e.preventDefault();
            e.stopPropagation();
            me.create_follow_up($(this).attr('data-name'));
        });
    },

    refresh: function() {
        var me = frappe.crm_dashboard;
        frappe.call({
            method: 'innomat.innomat.page.crm_dashboard.crm_dashboard.get_upcoming',
            args: { user: me.user_field ? me.user_field.get_value() : null },
            callback: function(r) {
                if (!r.message) { return; }
                me.render(r.message);
            }
        });
    },

    render: function(data) {
        var me = frappe.crm_dashboard;
        me.entries = {};
        (["Telefon", "Remotebesprechung", "Vor Ort Besprechung"]).forEach(function(type) {
            (data[type] || []).forEach(function(e) { me.entries[e.name] = e; });
        });
        var html = '<div class="crm-dashboard-frame"><div class="row">'
            + '<div class="col-sm-4 crm-dashboard-column">' + me.render_section(__("Nächste Telefon"), data["Telefon"] || []) + '</div>'
            + '<div class="col-sm-4 crm-dashboard-column">' + me.render_section(__("Nächste Remotebesprechungen"), data["Remotebesprechung"] || []) + '</div>'
            + '<div class="col-sm-4 crm-dashboard-column">' + me.render_section(__("Nächste Vor Ort Besprechungen"), data["Vor Ort Besprechung"] || []) + '</div>'
            + '</div></div>';
        me.body.html(html);
    },

    render_section: function(title, entries) {
        var html = '<div class="crm-dashboard-section"><h4 class="crm-dashboard-section-title">' + title + ' <span class="text-muted">(' + entries.length + ')</span></h4>';
        if (!entries.length) {
            return html + '<p class="text-muted crm-empty-state">' + __("Keine offenen Einträge") + '</p></div>';
        }
        html += '<ul class="list-unstyled crm-entry-list">';
        entries.forEach(function(e) {
            var subject = frappe.utils.escape_html(e.lead_name || e.company_name || e.lead || '');
            var when = frappe.datetime.str_to_user(e.date);
            if (e.time) { when += ' ' + e.time.substring(0, 5); }
            var overdue = e.date && e.date < frappe.datetime.get_today();
            html += '<li class="crm-entry' + (overdue ? ' crm-entry-overdue' : '') + '" data-name="' + frappe.utils.escape_html(e.name) + '">'
                + '<div class="crm-entry-header"><span class="crm-entry-date">' + when + '</span>'
                + ' <b class="crm-entry-subject">' + subject + '</b></div>'
                + (e.user ? ' <span class="text-muted">(' + frappe.utils.escape_html(e.user) + ')</span>' : '')
                + '<div class="crm-entry-actions">'
                + '<a href="#" class="crm-open-lead" data-lead="' + frappe.utils.escape_html(e.lead) + '">'
                + __("Lead öffnen") + '</a>'
                + '<a href="#" class="crm-create-follow-up" data-name="' + frappe.utils.escape_html(e.name) + '">'
                + __("Neue Kommunikation") + '</a></div>'
                + (e.note ? '<div class="text-muted small">' + frappe.utils.escape_html(e.note) + '</div>' : '')
                + (e.preparation ? '<div class="crm-entry-preparation"><span class="crm-entry-label">' + __("Vorbereitung") + ':</span> ' + frappe.utils.escape_html(e.preparation) + '</div>' : '')
                + '</li>';
        });
        html += '</ul>';
        return html + '</div>';
    },

    edit_entry: function(name) {
        var me = frappe.crm_dashboard;
        var doc = me.entries[name];
        if (!doc) { return; }
        var d = new frappe.ui.Dialog({
            title: __("Termin bearbeiten"),
            fields: [
                { fieldname: 'communication_type', label: __("Art"), fieldtype: 'Select',
                  options: "Telefon\nRemotebesprechung\nVor Ort Besprechung", reqd: 1, default: doc.communication_type },
                { fieldname: 'completed', label: __("Erledigt"), fieldtype: 'Check', default: 0 },
                { fieldname: 'cb', fieldtype: 'Column Break' },
                { fieldname: 'date', label: __("Datum"), fieldtype: 'Date', reqd: 1, default: doc.date },
                { fieldname: 'time', label: __("Zeit"), fieldtype: 'Time', default: doc.time },
                                { fieldname: 'duration', label: __("Dauer (Stunden)"), fieldtype: 'Float',
                                    default: doc.duration || 1, min: 0.25, precision: 2 },
                { fieldname: 'user', label: __("Benutzer"), fieldtype: 'Link', options: 'User', default: doc.user },
                { fieldname: 'sb', fieldtype: 'Section Break' },
                { fieldname: 'preparation', label: __("Vorbereitung"), fieldtype: 'Small Text', default: doc.preparation },
                { fieldname: 'note', label: __("Gesprächsnotiz"), fieldtype: 'Small Text', default: doc.note },
                { fieldname: 'follow_up', label: __("Nachbereitung"), fieldtype: 'Small Text', default: doc.follow_up }
            ],
            primary_action_label: __("Speichern"),
            primary_action: function(values) {
                frappe.call({
                    method: 'innomat.innomat.page.crm_dashboard.crm_dashboard.update_entry',
                    args: { name: name, values: values },
                    freeze: true,
                    callback: function(r) {
                        if (!r.exc) {
                            d.hide();
                            frappe.show_alert({ message: __("Gespeichert"), indicator: 'green' });
                            me.refresh();
                        }
                    }
                });
            }
        });
        d.show();
        },

        create_follow_up: function(name) {
            var me = frappe.crm_dashboard;
            var doc = me.entries[name];
            if (!doc) { return; }
            var d = new frappe.ui.Dialog({
                title: __("Neue Kommunikation für {0}", [doc.lead_name || doc.company_name || doc.lead]),
                fields: [
                    { fieldname: 'communication_type', label: __("Art"), fieldtype: 'Select',
                        options: "Telefon\nRemotebesprechung\nVor Ort Besprechung", reqd: 1, default: doc.communication_type || "Telefon" },
                    { fieldname: 'date', label: __("Datum"), fieldtype: 'Date', reqd: 1,
                        default: frappe.datetime.get_today() },
                    { fieldname: 'time', label: __("Zeit"), fieldtype: 'Time' },
                    { fieldname: 'duration', label: __("Dauer (Stunden)"), fieldtype: 'Float',
                        default: doc.duration || 1, min: 0.25, precision: 2 },
                    { fieldname: 'user', label: __("Benutzer"), fieldtype: 'Link', options: 'User',
                        default: doc.user || frappe.session.user },
                    { fieldname: 'sb', fieldtype: 'Section Break' },
                    { fieldname: 'preparation', label: __("Vorbereitung"), fieldtype: 'Small Text' },
                    { fieldname: 'note', label: __("Gesprächsnotiz"), fieldtype: 'Small Text' },
                    { fieldname: 'follow_up', label: __("Nachbereitung"), fieldtype: 'Small Text' }
                ],
                primary_action_label: __("Anlegen"),
                primary_action: function(values) {
                    frappe.call({
                        method: 'innomat.innomat.page.crm_dashboard.crm_dashboard.create_follow_up_entry',
                        args: { name: name, values: values },
                        freeze: true,
                        callback: function(r) {
                            if (!r.exc) {
                                d.hide();
                                frappe.show_alert({ message: __("Eintrag angelegt und alter erledigt"), indicator: 'green' });
                                me.refresh();
                            }
                        }
                    });
                }
            });
            d.show();
        },

        new_entry: function() {
                var me = frappe.crm_dashboard;
                var d = new frappe.ui.Dialog({
                        title: __("Neuer CRM-Eintrag"),
                        fields: [
                                { fieldname: 'communication_type', label: __("Art"), fieldtype: 'Select',
                                    options: "Telefon\nRemotebesprechung\nVor Ort Besprechung", reqd: 1, default: "Telefon" },
                                { fieldname: 'lead', label: __("Bestehender Lead"), fieldtype: 'Link',
                                    options: 'Lead' },
                                { fieldname: 'new_lead_name', label: __("Neuer Lead"), fieldtype: 'Data',
                                    description: __("Nur ausfüllen, wenn kein bestehender Lead ausgewählt ist") },
                                { fieldname: 'cb', fieldtype: 'Column Break' },
                                { fieldname: 'date', label: __("Datum"), fieldtype: 'Date', reqd: 1,
                                    default: frappe.datetime.get_today() },
                                { fieldname: 'time', label: __("Zeit"), fieldtype: 'Time' },
                                                                { fieldname: 'duration', label: __("Dauer (Stunden)"), fieldtype: 'Float',
                                                                    default: 1, min: 0.25, precision: 2 },
                                { fieldname: 'user', label: __("Benutzer"), fieldtype: 'Link', options: 'User',
                                    default: frappe.session.user },
                                { fieldname: 'sb', fieldtype: 'Section Break' },
                                { fieldname: 'preparation', label: __("Vorbereitung"), fieldtype: 'Small Text' },
                                { fieldname: 'note', label: __("Gesprächsnotiz"), fieldtype: 'Small Text' },
                                { fieldname: 'follow_up', label: __("Nachbereitung"), fieldtype: 'Small Text' }
                        ],
                        primary_action_label: __("Anlegen"),
                        primary_action: function(values) {
                                if (!values.lead && !values.new_lead_name) {
                                        frappe.msgprint(__("Bitte einen bestehenden Lead auswählen oder einen neuen Namen eingeben"));
                                        return;
                                }
                                frappe.call({
                                        method: 'innomat.innomat.page.crm_dashboard.crm_dashboard.create_entry',
                                        args: { values: values },
                                        freeze: true,
                                        callback: function(r) {
                                                if (!r.exc) {
                                                        d.hide();
                                                        frappe.show_alert({ message: __("Eintrag angelegt"), indicator: 'green' });
                                                        me.refresh();
                                                }
                                        }
                                });
                        }
                });
                d.show();
    }
};
